packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php

```diff
@@ -39,90 +44,123 @@ public function __construct(EventDispatcherInterface $eventDispatcher)
-        $mergedEntityUpdates = \array_merge(
-            $this->entityUpdates[$transactionNestingLevel] ?? [],
-            $entityUpdates
-        );
-        $this->entityUpdates[$transactionNestingLevel] = $mergedEntityUpdates;
+        $oid = \spl_object_hash($object);
+        $this->entityUpdates[$oid] = $object;
+        $this->entityChangeSets[$transactionNestingLevel][$oid] = $entityChangeSet;
     }
 
     public function disable(): void
     {
         $this->enabled = false;
+        $this->clear();
     }
 
     public function dispatch(): void
     {
-        $entityInsertions = $this->entityInsertions;
-        $entityUpdates = $this->entityUpdates;
-
-        $this->clear();
-
         if ($this->enabled === false) {
             return;
         }
 
-        $processedEntities = [];
-        foreach ($entityInsertions as $entities) {
-            foreach ($entities as $oid => $entity) {
-                $processedEntities[$oid] = $entity;
-                $this->eventDispatcher->dispatch(new EntityCreatedEvent($entity));
+        try {
+            $mergedEntityChangeSets = [];
+            foreach ($this->entityChangeSets as $levelChangeSets) {
+                foreach ($levelChangeSets as $oid => $changeSet) {
+                    $mergedEntityChangeSets[$oid] = $this->mergeChangeSet(
+                        $mergedEntityChangeSets[$oid] ?? [],
+                        $changeSet
+                    );
+                }
             }
-        }
 
-        foreach ($entityUpdates as $entities) {
-            foreach ($entities as $oid => $entity) {
-                if (isset($processedEntities[$oid])) {
-                    continue;
-                }
-                $processedEntities[$oid] = $entity;
-                $this->eventDispatcher->dispatch(new EntityUpdatedEvent($entity));
+            foreach ($mergedEntityChangeSets as $oid => $entityChangeSet) {
+                $event = $this->createEntityEvent($oid, $entityChangeSet);
+
+                $this->eventDispatcher->dispatch($event);
             }
+        } finally {
+            $this->clear();
         }
     }
 
     public function enable(): void
     {
         $this->enabled = true;
     }
+
+    /**
+     * @param string $oid
+     * @param array<string, array{mixed, mixed}> $entityChangeSet
+     *
+     * @return \EonX\EasyCore\Doctrine\Events\EntityCreatedEvent|\EonX\EasyCore\Doctrine\Events\EntityUpdatedEvent
+     */
+    private function createEntityEvent(string $oid, array $entityChangeSet)
+    {
+        if (isset($this->entityInsertions[$oid]) !== false) {
+            return new EntityCreatedEvent($this->entityInsertions[$oid], $entityChangeSet);
+        }
+
+        if (isset($this->entityUpdates[$oid]) !== false) {
+            return new EntityUpdatedEvent($this->entityUpdates[$oid], $entityChangeSet);
+        }
+
+        // @codeCoverageIgnoreStart
+        throw new \LogicException('Entity fot event not found');
```

Comment:
Please fix the typo and add a point at the end of the message.

Reply:
Fixed.